### PR TITLE
fix race between chunk transactions and schema changes (i.e. keep locks for follow-up chunks)

### PR DESCRIPTION
**Pull request by dorinhogea**
Currently we drop all the locks when a transaction is committed.  Therefore the second transaction chunk has no locks on used tables, and a race with a schema change on those tables has unfortunate consequences. 
For delicate reasons it is not possible to keep locks across chunks.  Solution is to reacquire locks when a new chunk is created.
NOTE: in order to make sure the table has not changed while the transaction lost the table locks, we need to cache their versions in sqlite (currently, only remote table versions are cached, but this patch adds the version for local tables as well).

Signed-off-by: Dorin Hogea <dhogea@bloomberg.net>

### CODE REVIEW

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>1/410 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>tunables</p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
```

File: db/sqlglue.c [hunk 1]
```diff
@@ -8439,6 +8482,27 @@ static int chunk_transaction(BtCursor *pCur, struct sqlclntstate *clnt,
...
+            rdlock_schema_lk();
+            assert(pCur->vdbe == (Vdbe*)clnt->dbtran.pStmt);
+            newlocks_rc = sqlite3LockStmtTables(clnt->dbtran.pStmt);
+            unlock_schema_lk();
+
+            if (newlocks_rc) {
+                comdb2_sqlite3VdbeError(
+                        pCur->vdbe,
+                        "Failed to renew locks, table schema changed");
+                rc = SQLITE_ERROR;
```
Comment (nirbhayc) [inline db/sqlglue.c:8502]: rc = newlocks_rc?

File: db/sqlglue.c [hunk 2]
```diff
@@ -8412,6 +8451,10 @@ static int chunk_transaction(BtCursor *pCur, struct sqlclntstate *clnt,
              */
         }
 
+        if (gbl_throttle_txn_chunks_sec > 0) {
+            sleep(gbl_throttle_txn_chunks_sec);
```
Comment (mponomar) [inline db/sqlglue.c:8455]: Make this milliseconds?

**Review by mponomar — APPROVED** [review]

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: Success &#10003;.</p>

**Pull request merged**